### PR TITLE
feat(mobile): edit wheel manually bb-344

### PR DESCRIPTION
**Pull request by Olha-Katran**
Create scores editing form using slider with gradient color
Add nested route from "wheel" to "Edit my wheel results"
Add style to page "Wheel"
Add  checkbox form to "Retake Quiz" tabs
Add getScore and editScore slice
Create subheader -  switch between "Edit scores manually" forms and "retake quiz".
 I didn't add horizontal scrolling because the sliders don't work correctly with it.
 
 
 
screenshot:
   

Video: 

https://github.com/user-attachments/assets/24b01bec-d2b9-4c9e-bcfe-87b6f15cd2c2

### CODE REVIEW

**Comment by anogire** [comment]
We  need to output all categories with previous Quiz values. The video does not show these categories.
Screenshot 2024-09-16 at 22 42 37

File: apps/mobile/src/screens/edit-wheel-results/styles.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+import { BaseColor } from "~/libs/enums/enums";
+
+const styles = StyleSheet.create({
+	checkboxForm: {
+		borderColor: BaseColor.LIGHT_GRAY,
+		borderRadius: 14,
+		borderWidth: 1,
+		height: 500,
+	},
```
Comment (anogire) [inline apps/mobile/src/screens/edit-wheel-results/styles.ts:11]: Do we need that definition of height?
Can it be 100%?

File: apps/mobile/src/screens/edit-wheel-results/styles.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+	container: {
+		marginBottom: 0,
+		paddingBottom: 20,
+	},
```
Comment (anogire) [inline apps/mobile/src/screens/edit-wheel-results/styles.ts:15]: Can it be pb24 from globalStyles?

File: apps/mobile/src/libs/helpers/get-formatted-date/get-formatted-date.helper.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import { format } from "date-fns";
```
Comment (fshabanov) [inline apps/mobile/src/libs/helpers/get-formatted-date/get-formatted-date.helper.ts:1]: Move this to shared

File: apps/mobile/src/libs/components/gradient-slider/gradient-slider.tsx [hunk 4]
```diff
@@ -113,15 +123,16 @@ const GradientSlider: React.FC<Properties> = ({
 				/>
 
 				<Slider
+					hitSlop={{ bottom: 20, left: 20, right: 20, top: 20 }}
```
Comment (fshabanov) [inline apps/mobile/src/libs/components/gradient-slider/gradient-slider.tsx:126]: Avoid magic numbers

File: apps/mobile/src/libs/components/multiple-slider-input/multiple-slider-input.tsx [hunk 5]
```diff
@@ -0,0 +1,62 @@
+import React, { useCallback } from "react";
```
Comment (fshabanov) [inline apps/mobile/src/libs/components/multiple-slider-input/multiple-slider-input.tsx:1]: Import hooks from ~/libs/hooks/hooks

File: apps/mobile/src/libs/components/multiple-slider-input/multiple-slider-input.tsx [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+type Properties<T extends FieldValues> = {
+	control: Control<T, null>;
+	data: SliderData[];
+	name: FieldPath<T>;
+};
+
+type ScoreItem = { categoryId: number; score: number };
+
+const NEGATIVE_INDEX = -1;
```
Comment (fshabanov) [inline apps/mobile/src/libs/components/multiple-slider-input/multiple-slider-input.tsx:20]:
```suggestion
const NO_ITEM_INDEX = -1;
```

File: apps/mobile/src/libs/components/multiple-slider-input/multiple-slider-input.tsx [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+						);
+
+			onChange(updatedItems);
+		},
+		[value, onChange],
+	);
+
+	return (
+		<View>
+			{data.map((item) => (
```
Comment (fshabanov) [inline apps/mobile/src/libs/components/multiple-slider-input/multiple-slider-input.tsx:51]: B27

File: apps/mobile/src/libs/components/page-switcher/page-switcher.tsx [hunk 8]
```diff
@@ -0,0 +1,74 @@
+import React, { useCallback } from "react";
```
Comment (fshabanov) [inline apps/mobile/src/libs/components/page-switcher/page-switcher.tsx:1]: Same about importing hooks

File: apps/mobile/src/libs/components/page-switcher/page-switcher.tsx [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+		[onTabChange],
+	);
+
+	const handlePress = (tab: string): (() => void) => {
+		return () => {
+			handleTabPress(tab);
+		};
+	};
+
+	const renderedTabs = tabs.map((tab) => {
```
Comment (fshabanov) [inline apps/mobile/src/libs/components/page-switcher/page-switcher.tsx:33]: A10

File: apps/mobile/src/libs/components/page-switcher/style.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
+import { StyleSheet } from "react-native";
+
+import { BaseColor } from "~/libs/enums/app/base-color.enum";
```
Comment (fshabanov) [inline apps/mobile/src/libs/components/page-switcher/style.ts:3]:
```suggestion
import { BaseColor } from "~/libs/enums/enums";
```

File: apps/mobile/src/screens/edit-wheel-results/edit-wheel-results.tsx [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+	useNavigation,
+	useState,
+} from "~/libs/hooks/hooks";
+import { globalStyles } from "~/libs/styles/styles";
+import {
+	type NativeStackNavigationProp,
+	type RootNavigationParameterList,
+} from "~/libs/types/types";
+import { type QuizScoresUpdateRequestDto } from "~/packages/quiz/quiz";
+import { ScoresEditForm } from "~/screens/edit-wheel-results/libs/components/scores-edit-form/scores-edit-form";
```
Comment (fshabanov) [inline apps/mobile/src/screens/edit-wheel-results/edit-wheel-results.tsx:27]:
```suggestion
import { ScoresEditForm } from "./libs/components/components";
```

File: apps/mobile/src/screens/edit-wheel-results/edit-wheel-results.tsx [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+		);
+	}, []);
+
+	return (
+		<ScreenWrapper
+			edges={["top", "left", "right"]}
+			style={{ backgroundColor: BaseColor.BG_WHITE }}
+		>
+			<LoaderWrapper isLoading={dataStatus === DataStatus.PENDING}>
+				<View style={[styles.container]}>
```
Comment (fshabanov) [inline apps/mobile/src/screens/edit-wheel-results/edit-wheel-results.tsx:65]:
```suggestion
				<View style={styles.container}>
```

File: apps/mobile/src/screens/edit-wheel-results/edit-wheel-results.tsx [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+	const handleTabChange = useCallback(() => {
+		setActiveTab((previousTab) =>
+			previousTab === "Edit manually" ? "Retake Quiz" : "Edit manually",
+		);
+	}, []);
+
+	return (
+		<ScreenWrapper
+			edges={["top", "left", "right"]}
+			style={{ backgroundColor: BaseColor.BG_WHITE }}
```
Comment (fshabanov) [inline apps/mobile/src/screens/edit-wheel-results/edit-wheel-results.tsx:62]: Avoid inline styles

File: apps/mobile/src/screens/edit-wheel-results/libs/components/scores-edit-form/scores-edit-form.tsx [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+import { BottomTabScreenName } from "~/libs/enums/enums";
+import { useAppForm, useCallback, useNavigation } from "~/libs/hooks/hooks";
+import { globalStyles } from "~/libs/styles/styles";
+import { type BottomTabNavigationParameterList } from "~/libs/types/types";
+import {
+	type QuizScoresGetAllItemResponseDto,
+	type QuizScoresUpdateRequestDto,
+	updateScoresValidationSchema,
+} from "~/packages/quiz/quiz";
+import { SCORE_FORM_DEFAULT_VALUES } from "~/screens/edit-wheel-results/libs/constants/constants";
```
Comment (fshabanov) [inline apps/mobile/src/screens/edit-wheel-results/libs/components/scores-edit-form/scores-edit-form.tsx:19]:
```suggestion
import { SCORE_FORM_DEFAULT_VALUES } from "../../constants/constants";
```

File: apps/mobile/src/screens/wheel/wheel.tsx [hunk 15]
```diff
@@ -1,21 +1,64 @@
...
+		void dispatch(quizActions.getScores());
+	}, [dispatch]);
+
 	return (
-		<ScreenWrapper>
-			<Text>My wheel results</Text>
-			<View style={globalStyles.alignItemsCenter}>
-				<WheelChart size={WHEEL_SIZE} />
+		<ScreenWrapper style={{ backgroundColor: BaseColor.BG_WHITE }}>
+			<View style={[globalStyles.m16]}>
```
Comment (fshabanov) [inline apps/mobile/src/screens/wheel/wheel.tsx:44]:
```suggestion
			<View style={globalStyles.m16}>
```

File: apps/mobile/src/screens/wheel/wheel.tsx [hunk 16]
```diff
@@ -1,21 +1,64 @@
...
+	useEffect(() => {
+		void dispatch(quizActions.getScores());
+	}, [dispatch]);
+
 	return (
-		<ScreenWrapper>
-			<Text>My wheel results</Text>
-			<View style={globalStyles.alignItemsCenter}>
-				<WheelChart size={WHEEL_SIZE} />
+		<ScreenWrapper style={{ backgroundColor: BaseColor.BG_WHITE }}>
```
Comment (fshabanov) [inline apps/mobile/src/screens/wheel/wheel.tsx:43]: Avoid inline styles

File: apps/mobile/src/screens/wheel/wheel.tsx [hunk 17]
```diff
@@ -1,21 +1,64 @@
...
+					preset="tag"
+					style={[globalStyles.pv4, globalStyles.ph12, styles.date]}
+				>
+					{date}
+				</Text>
+				<View style={styles.container}>
+					<View style={globalStyles.alignItemsCenter}>
+						<WheelChart size={WHEEL_SIZE} />
+					</View>
+					<View style={[globalStyles.mh48]}>
```
Comment (fshabanov) [inline apps/mobile/src/screens/wheel/wheel.tsx:58]:
```suggestion
					<View style={globalStyles.mh48}>
```

File: apps/mobile/src/slices/quiz/actions.ts [hunk 18]
```diff
@@ -28,4 +30,14 @@ const getAllQuestions = createAsyncThunk<
 	return await quizApi.getAllQuestions();
 });
 
-export { getAllQuestions, getScores };
+const getScores = createAsyncThunk<
+	QuizScoresGetAllResponseDto,
+	undefined,
+	AsyncThunkConfig
+>(`${sliceName}/score`, async (_, { extra }) => {
```
Comment (fshabanov) [inline apps/mobile/src/slices/quiz/actions.ts:37]:
```suggestion
>(`${sliceName}/get-scores`, async (_, { extra }) => {
```

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/mobile/src/libs/components/multiple-slider-input/multiple-slider-input.tsx [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+	control,
+	data,
+	name,
+}: Properties<T>): JSX.Element | null => {
+	const { field } = useFormController({ control, name });
+	const { onChange, value } = field;
+
+	const handleSliderChange = useCallback(
+		(categoryId: number, score: number) => {
+			const currentItems: ScoreItem[] = value || [];
```
Comment (fshabanov) [inline apps/mobile/src/libs/components/multiple-slider-input/multiple-slider-input.tsx:32]:
```suggestion
			const currentItems: ScoreItem[] = value ?? [];
```

File: apps/mobile/src/libs/components/slider-content/slider-content.tsx [hunk 20]
```diff
@@ -1,34 +1,44 @@
...
 };
 
 const MAX_SLIDER_VALUE = 10;
-const MIN_SLIDER_VALUE = 0;
 
-const SliderContent: React.FC<Properties> = ({ color, label }) => {
+const getGradientColorsForCategory = (
+	categoryName: string,
+): keyof typeof colorToGradientColors => {
+	return categoryToColors[categoryName.replaceAll(/\s+/g, "")] ?? "blue";
```
Comment (fshabanov) [inline apps/mobile/src/libs/components/slider-content/slider-content.tsx:18]: Move "blue" to a constant, it's not clear why blue

File: packages/shared/src/libs/helpers/get-formatted-date/get-formatted-date.helper.ts [hunk 21]
```diff
@@ -0,0 +1,9 @@
+import { format } from "date-fns";
+
+const getFormattedDate = (): string => {
```
Comment (fshabanov) [inline packages/shared/src/libs/helpers/get-formatted-date/get-formatted-date.helper.ts:3]: Send format as an argument

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: packages/shared/src/libs/helpers/get-formatted-date/get-formatted-date.helper.ts (hunk 21 above)
Comment (Olha-Katran) [reply packages/shared/src/libs/helpers/get-formatted-date/get-formatted-date.helper.ts:3]: Should I change the location of the "date-fns" package? (It's currently in dev dependencies in shared)

File: apps/mobile/src/libs/components/slider-content/slider-content.tsx (hunk 20 above)
Comment (Olha-Katran) [reply apps/mobile/src/libs/components/slider-content/slider-content.tsx:18]: delete unnecessary condition

File: packages/shared/src/libs/helpers/get-formatted-date/get-formatted-date.helper.ts (hunk 21 above)
Comment (fshabanov) [reply packages/shared/src/libs/helpers/get-formatted-date/get-formatted-date.helper.ts]: It is not a dev dependency, we need this package in production. Should be in dependencies in shared

File: apps/mobile/src/libs/components/page-switcher/page-switcher.tsx [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+					globalStyles.pv8,
+					styles.tab,
+					isActive ? styles.activeTab : styles.inactiveTab,
+				]}
+			>
+				<Text
+					preset="regular"
+					style={{
+						color: isActive ? BaseColor.BLACK : BaseColor.GRAY,
+					}}
```
Comment (fshabanov) [inline apps/mobile/src/libs/components/page-switcher/page-switcher.tsx:53]:
```suggestion
					style={isActive ? styles.active : styles.inactive}
```
Same approach as for TouchableOpacity

File: apps/mobile/src/screens/edit-wheel-results/edit-wheel-results.tsx [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+	const dispatch = useAppDispatch();
+
+	const handleEditScores = useCallback(
+		(payload: QuizScoresUpdateRequestDto) => {
+			void dispatch(quizActions.editScores(payload));
+		},
+		[dispatch],
+	);
+
+	const [activeTab, setActiveTab] = useState<string>("Edit manually");
```
Comment (fshabanov) [inline apps/mobile/src/screens/edit-wheel-results/edit-wheel-results.tsx:41]:
```suggestion
	const [activeTab, setActiveTab] = useState<ValueOf<typeof Enum>>("Edit manually");
```
Use enum for tab

File: packages/shared/src/libs/helpers/get-formatted-date/get-formatted-date.helper.ts [hunk 24]
```diff
@@ -0,0 +1,9 @@
+import { format } from "date-fns";
+
+const getFormattedDate = (dateFormat: string): string => {
```
Comment (fshabanov) [inline packages/shared/src/libs/helpers/get-formatted-date/get-formatted-date.helper.ts:3]:
```suggestion
type DateFormat = "d MMM yyyy, EEEE";

const getFormattedDate = (dateFormat: DateFormat): string => {
```

File: packages/shared/package.json [hunk 25]
```diff
@@ -15,6 +15,7 @@
 		"build": "tsc -p tsconfig.build.json && tsc-alias -p tsconfig.build.json"
 	},
 	"devDependencies": {
-		"zod": "3.23.8"
+		"zod": "3.23.8",
+		"date-fns": "3.6.0"
```
Comment (fshabanov) [inline packages/shared/package.json:19]: Move to dependencies. As you do that, move zod as well to dependencies, please

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: packages/shared/package.json (hunk 25 above)
Comment (Olha-Katran) [reply packages/shared/package.json:19]: Check it, please

File: apps/mobile/src/screens/edit-wheel-results/edit-wheel-results.tsx [hunk 26]
```diff
@@ -0,0 +1,137 @@
...
+import { type QuizScoresUpdateRequestDto } from "~/packages/quiz/quiz";
+import { actions as quizActions } from "~/slices/quiz/quiz";
+
+import { ScoresEditForm } from "./libs/components/componets";
+import { styles } from "./styles";
+
+const EditWheelResultsTabs = {
+	EDIT_MANUALLY: "Edit manually",
+	RETAKE_QUIZ: "Retake Quiz",
+};
```
Comment (fshabanov) [inline apps/mobile/src/screens/edit-wheel-results/edit-wheel-results.tsx:33]:
```suggestion
const EditWheelResultsTab = {
	EDIT_MANUALLY: "Edit manually",
	RETAKE_QUIZ: "Retake Quiz",
} as const;
```

File: apps/mobile/src/screens/edit-wheel-results/libs/components/scores-edit-form/scores-edit-form.tsx [hunk 27]
```diff
@@ -0,0 +1,73 @@
...
+	type QuizScoresGetAllItemResponseDto,
+	type QuizScoresUpdateRequestDto,
+	updateScoresValidationSchema,
+} from "~/packages/quiz/quiz";
+
+import { SCORE_FORM_DEFAULT_VALUES } from "../../constants/constants";
+import { styles } from "./styles";
+
+type Properties = {
+	data: QuizScoresGetAllItemResponseDto[];
```
Comment (fshabanov) [inline apps/mobile/src/screens/edit-wheel-results/libs/components/scores-edit-form/scores-edit-form.tsx:24]:
```suggestion
	scores: QuizScoresGetAllItemResponseDto[];
```
Let's be more specific

**Review by fshabanov — CHANGES_REQUESTED** [review]

**Review by revenets — APPROVED** [review]

File: apps/mobile/src/libs/components/slider-content/slider-content.tsx [hunk 28]
```diff
@@ -1,34 +1,47 @@
...
+	score: SliderData;
+};
+
+const getGradientColorsForCategory = (
+	categoryName: string,
+): keyof typeof colorToGradientColors => {
+	return (
+		categoryToColors[categoryName.replaceAll(/\s+/g, "")] ??
+		DEFAULT_SLIDER_COLOR
+	);
```
Comment (fshabanov) [inline apps/mobile/src/libs/components/slider-content/slider-content.tsx:21]: Move to a separate helper file

File: apps/mobile/src/screens/edit-wheel-results/edit-wheel-results.tsx [hunk 29]
```diff
@@ -0,0 +1,133 @@
...
+	const [activeTab, setActiveTab] = useState<
+		ValueOf<typeof EditWheelResultsTab>
+	>(EditWheelResultsTab.EDIT_MANUALLY);
+
+	const handleTabChange = useCallback(() => {
+		setActiveTab((previousTab) =>
+			previousTab === EditWheelResultsTab.EDIT_MANUALLY
+				? EditWheelResultsTab.RETAKE_QUIZ
+				: EditWheelResultsTab.EDIT_MANUALLY,
+		);
```
Comment (fshabanov) [inline apps/mobile/src/screens/edit-wheel-results/edit-wheel-results.tsx:54]:
```suggestion
		setActiveTab((previousTab) => {
			return previousTab === EditWheelResultsTab.EDIT_MANUALLY
				? EditWheelResultsTab.RETAKE_QUIZ
				: EditWheelResultsTab.EDIT_MANUALLY,
		});
```

**Review by fshabanov — APPROVED** [review]

**Pull request merged**